dlc/src/lib.rs

```diff
@@ -0,0 +1,703 @@
+}
+
+impl InputType {
+    fn get_size(&self) -> usize {
+        match self {
+            InputType::P2WPKH => P2WPKH_SPK_SIZE * 4 + P2WPKH_WITNESS_SIZE,
+        }
+    }
+}
+
+/// Contains info about a utxo used for funding a DLC contract
+pub struct TxInputInfo {
+    /// The outpoint for the utxo
+    pub outpoint: OutPoint,
+    /// The type of input
+    pub input_type: InputType,
+}
+
+/// An error code.
+#[derive(Copy, PartialEq, Eq, Clone, Debug)]
+pub enum Error {
+    /// Secp256k1 error
+    Base(secp256k1::Error),
+    /// An invalid argument was provided
+    InvalidArgument,
+}
+
+impl From<secp256k1::Error> for Error {
+    fn from(error: secp256k1::Error) -> Error {
+        Error::Base(error)
+    }
+}
+
+/// Contains the parameters required for creating DLC transactions for a single
+/// party. Specifically these are the common fields between Offer and Accept
+/// messages.
+pub struct PartyParams {
+    /// The public key for the fund multisig script
+    pub fund_pubkey: PublicKey,
+    /// An address to receive change
+    pub change_address: Address,
+    /// An address to receive the outcome amount
+    pub final_address: Address,
+    /// A list of input to fund the contract
+    pub inputs: Vec<TxInputInfo>,
+    /// The total value of the provided input
+    pub input_amount: u64,
+    /// The collateral put in the contract by the party
+    pub collateral: u64,
+}
+
+impl PartyParams {
+    fn get_change_output_and_fees(&self, fee_rate: u64) -> Result<(TxOut, u64, u64), Error> {
+        let input_size: usize = self.inputs.iter().map(|i| i.input_type.get_size()).sum();
+        let change_size = self.change_address.script_pubkey().len();
+        let fund_fee = u64::try_from(FUND_TX_BASE_WEIGHT / 2 + input_size + change_size * 4)
+            .unwrap()
+            * fee_rate;
+        let cet_fee = u64::try_from(CET_BASE_WEIGHT + self.final_address.script_pubkey().len() * 4)
+            .unwrap()
+            * fee_rate;
+        let fund_out = self.collateral + fund_fee + cet_fee;
+        if self.input_amount < fund_out {
+            return Err(Error::InvalidArgument);
+        }
+
+        let change_output = TxOut {
+            value: self.input_amount - fund_out,
+            script_pubkey: self.change_address.script_pubkey(),
+        };
+
+        Ok((change_output, fund_fee, cet_fee))
+    }
+
+    fn get_tx_input_info(&self, sequence: u32) -> Vec<TxIn> {
+        self.inputs
+            .iter()
+            .map(|i| TxIn {
+                previous_output: i.outpoint,
+                script_sig: Script::new(),
+                sequence,
+                witness: Vec::new(),
+            })
+            .collect()
+    }
+}
+
+/// Create the transactions for a DLC contract based on the provided parameters
+pub fn create_dlc_transactions(
+    local_params: &PartyParams,
+    remote_params: &PartyParams,
+    outcomes: &[Outcome],
+    refund_lock_time: u32,
+    fee_rate: u64,
+    lock_time: u32,
```

Comment:
it is also applied to the fund transaction. I feel `fund_and_cet_lock_time` is a bit too long, but if you like it I can change it to that.

Reply:
I don't think this should be applied to the funding tx in the current spec?

Reply:
As discussed we need to update the specs to include fund tx lock time and cet lock time to enable anti-fee snipping mechanisms. I divided the previous `lock_time` into `fund_lock_time` and `cet_lock_time` to avoid fingerprinting.

Reply:
Yes but they should be different locktime, CET's one have still a maturing bound ?

Reply:
They are now different, we gave up on setting locktime of CET to contract maturity because it doesn't provide any big security advantage.